Error building with ghc-8.10.2

[tcdName](https://github.com/haskell/haddock/blob/ghc-8.10/haddock-api/src/Haddock/GhcUtils.hs#L62) have different type in [ghc-8.10.2](https://downloads.haskell.org/~ghc/8.10.2/docs/html/libraries/ghc-8.10.2/GHC-Hs-Decls.html#v:tcdName) (compare to [ghc-8.10.1](https://downloads.haskell.org/~ghc/8.10.1/docs/html/libraries/ghc-8.10.1/GHC-Hs-Decls.html#v:tcdName))

```haskell
tcdName :: TyClDecl pass -> IdP pass  -- ghc-8.10.1
tcdName :: TyClDecl (GhcPass p) -> IdP (GhcPass p)  -- ghc-8.10.2
```

> haddock-api> /tmp/stack-a8f062bc5e8dbbce/haddock-api-2.24.0/src/Haddock/GhcUtils.hs:60:42: error:
> haddock-api>     • Could not deduce: p ~ GhcPass p0
> haddock-api>       from the context: (SrcSpanLess (LPat p) ~ Pat p,
> haddock-api>                          HasSrcSpan (LPat p))
> haddock-api>         bound by the type signature for:
> haddock-api>                    getMainDeclBinder :: forall p.
> haddock-api>                                         (SrcSpanLess (LPat p) ~ Pat p, HasSrcSpan (LPat p)) =>
> haddock-api>                                         HsDecl p -> [IdP p]
> haddock-api>         at src/Haddock/GhcUtils.hs:(58,1)-(59,40)
> haddock-api>       ‘p’ is a rigid type variable bound by
> haddock-api>         the type signature for:
> haddock-api>           getMainDeclBinder :: forall p.
> haddock-api>                                (SrcSpanLess (LPat p) ~ Pat p, HasSrcSpan (LPat p)) =>
> haddock-api>                                HsDecl p -> [IdP p]
> haddock-api>         at src/Haddock/GhcUtils.hs:(58,1)-(59,40)
> haddock-api>       Expected type: TyClDecl (GhcPass p0)
> haddock-api>         Actual type: TyClDecl p
> haddock-api>     • In the first argument of ‘tcdName’, namely ‘d’
> haddock-api>       In the expression: tcdName d
> haddock-api>       In the expression: [tcdName d]
> haddock-api>     • Relevant bindings include
> haddock-api>         d :: TyClDecl p (bound at src/Haddock/GhcUtils.hs:60:28)
> haddock-api>         getMainDeclBinder :: HsDecl p -> [IdP p]
> haddock-api>           (bound at src/Haddock/GhcUtils.hs:60:1)
> haddock-api>    |
> haddock-api> 60 | getMainDeclBinder (TyClD _ d) = [tcdName d]
> haddock-api>    |                                          ^
> haddock-api>

--- a/haddock-api/src/Haddock/Backends/LaTeX.hs
+++ b/haddock-api/src/Haddock/Backends/LaTeX.hs
@@ -255,7 +255,7 @@ declNames :: LHsDecl DocNameI
              , [DocName]       --   names being declared
              )
 declNames (L _ decl) = case decl of
-  TyClD _ d  -> (empty, [tcdName d])
+  TyClD _ d  -> (empty, [tcdNameI d])
   SigD _ (TypeSig _ lnames _ ) -> (empty, map unLoc lnames)
   SigD _ (PatSynSig _ lnames _) -> (text "pattern", map unLoc lnames)
   ForD _ (ForeignImport _ (L _ n) _ _) -> (empty, [n])

--- a/haddock-api/src/Haddock/Backends/Xhtml.hs
+++ b/haddock-api/src/Haddock/Backends/Xhtml.hs
@@ -407,7 +407,7 @@ ppJsonIndex odir maybe_source_url maybe_wiki_url unicode pkg qual_opt ifaces = d
     exportSubs _ = []
 
     exportName :: ExportItem DocNameI -> [IdP DocNameI]
-    exportName ExportDecl { expItemDecl } = getMainDeclBinder (unLoc expItemDecl)
+    exportName ExportDecl { expItemDecl } = getMainDeclBinderI (unLoc expItemDecl)
     exportName ExportNoDecl { expItemName } = [expItemName]
     exportName _ = []
 

--- a/haddock-api/src/Haddock/Backends/Xhtml/Decl.hs
+++ b/haddock-api/src/Haddock/Backends/Xhtml/Decl.hs
@@ -536,6 +536,8 @@ ppClassDecl summary links instances fixities loc d subdocs
     -- Only the fixity relevant to the class header
     fixs = ppFixities [ f | f@(n,_) <- fixities, n == unLoc lname ] qual
 
+    nm   = tcdNameI decl
+
     hdr = ppClassHdr summary lctxt (unLoc lname) ltyvars lfds
 
     -- Associated types
@@ -794,7 +796,7 @@ ppDataDecl summary links instances fixities subdocs loc doc dataDecl pats
   | otherwise = header_ +++ docSection curname pkg qual doc +++ constrBit +++ patternBit +++ instancesBit
 
   where
-    docname   = tcdName dataDecl
+    docname   = tcdNameI dataDecl
     curname   = Just $ getName docname
     cons      = dd_cons (tcdDataDefn dataDecl)
     isH98     = case unLoc (head cons) of

--- a/haddock-api/src/Haddock/GhcUtils.hs
+++ b/haddock-api/src/Haddock/GhcUtils.hs
@@ -58,8 +58,7 @@ moduleString = moduleNameString . moduleName
 isNameSym :: Name -> Bool
 isNameSym = isSymOcc . nameOccName
 
-getMainDeclBinder :: (SrcSpanLess (LPat p) ~ Pat p , HasSrcSpan (LPat p)) =>
-                     HsDecl p -> [IdP p]
+getMainDeclBinder :: HsDecl (GhcPass p) -> [IdP (GhcPass p)]
 getMainDeclBinder (TyClD _ d) = [tcdName d]
 getMainDeclBinder (ValD _ d) =
   case collectHsBindBinders d of
@@ -220,6 +219,31 @@ getGADTConType (ConDeclGADT { con_forall = L _ has_forall
 getGADTConType (ConDeclH98 {}) = panic "getGADTConType"
   -- Should only be called on ConDeclGADT
 getGADTConType (XConDecl nec) = noExtCon nec
+
+getMainDeclBinderI :: HsDecl DocNameI -> [IdP DocNameI]
+getMainDeclBinderI (TyClD _ d) = [tcdNameI d]
+getMainDeclBinderI (ValD _ d) =
+  case collectHsBindBinders d of
+    []       -> []
+    (name:_) -> [name]
+getMainDeclBinderI (SigD _ d) = sigNameNoLoc d
+getMainDeclBinderI (ForD _ (ForeignImport _ name _ _)) = [unLoc name]
+getMainDeclBinderI (ForD _ (ForeignExport _ _ _ _)) = []
+getMainDeclBinderI _ = []
+
+familyDeclLNameI :: FamilyDecl DocNameI -> Located DocName
+familyDeclLNameI (FamilyDecl { fdLName = n }) = n
+familyDeclLNameI (XFamilyDecl nec) = noExtCon nec
+
+tyClDeclLNameI :: TyClDecl DocNameI -> Located DocName
+tyClDeclLNameI (FamDecl { tcdFam = fd })     = familyDeclLNameI fd
+tyClDeclLNameI (SynDecl { tcdLName = ln })   = ln
+tyClDeclLNameI (DataDecl { tcdLName = ln })  = ln
+tyClDeclLNameI (ClassDecl { tcdLName = ln }) = ln
+tyClDeclLNameI (XTyClDecl nec) = noExtCon nec
+
+tcdNameI :: TyClDecl DocNameI -> DocName
+tcdNameI = unLoc . tyClDeclLNameI
 
 -- -------------------------------------
 
@@ -761,4 +785,3 @@ defaultRuntimeRepVars = go emptyVarEnv
 
     go _ ty@(LitTy {}) = ty
     go _ ty@(CoercionTy {}) = ty
-

--- a/haddock-api/src/Haddock/Interface/Rename.hs
+++ b/haddock-api/src/Haddock/Interface/Rename.hs
@@ -29,6 +29,7 @@ import TysWiredIn (eqTyCon_RDR)
 import Control.Applicative
 import Control.Arrow ( first )
 import Control.Monad hiding (mapM)
+import Data.List (intercalate)
 import qualified Data.Map as Map hiding ( Map )
 import qualified Data.Set as Set
 import Prelude hiding (mapM)

--- a/haddock-api/src/Haddock/Types.hs
+++ b/haddock-api/src/Haddock/Types.hs
@@ -789,6 +789,7 @@ type instance XDataDecl     DocNameI = NoExtField
 type instance XSynDecl      DocNameI = NoExtField
 type instance XFamDecl      DocNameI = NoExtField
 type instance XXFamilyDecl  DocNameI = NoExtCon
+type instance XXTyClDecl    DocNameI = NoExtCon
 
 type instance XHsIB             DocNameI _ = NoExtField
 type instance XHsWC             DocNameI _ = NoExtField

